database/memory/db.go

```diff
@@ -535,34 +587,204 @@ func (db *MemoryDB) removeAllIndices(address types.Address) error {
 	return nil
 }
 
+func (db *MemoryDB) getERC20EntryAtBlock(contract types.Address, holder types.Address, block uint64) (*ERC20TokenHolder, error) {
+	db.mux.RLock()
+	defer db.mux.RUnlock()
+	var tmpItem *ERC20TokenHolder
+	found := false
+	for i, item := range db.erc20BalancesDB {
+		if item.BlockNumber <= block && item.Contract == contract && item.Holder == holder {
+			if !found {
+				tmpItem = &db.erc20BalancesDB[i]
+				found = true
+			} else {
+				if item.BlockNumber > tmpItem.BlockNumber {
+					tmpItem = &db.erc20BalancesDB[i]
+				}
+			}
+		}
+	}
+	if !found {
+		return nil, database.ErrNotFound
+	}
+	return tmpItem, nil
+}
+
 func (db *MemoryDB) RecordNewERC20Balance(contract types.Address, holder types.Address, block uint64, amount *big.Int) error {
+	existingTokenEntry, errExisting := db.getERC20EntryAtBlock(contract, holder, block-1)
+	db.mux.Lock()
+	defer db.mux.Unlock()
+	if errExisting != nil && errExisting != database.ErrNotFound {
+		return errExisting
+	}
+
+	//add new entry
+	tokenInfo := ERC20TokenHolder{
+		Contract:    contract,
+		Holder:      holder,
+		BlockNumber: block,
+		Amount:      amount.String(),
+	}
+	db.erc20BalancesDB = append(db.erc20BalancesDB, tokenInfo)
+	/////
+	if errExisting == database.ErrNotFound {
+		return nil
+	}
+	blk := block - 1
+	existingTokenEntry.HeldUntil = &blk
+
 	return nil
 }
 
 func (db *MemoryDB) GetERC20Balance(contract types.Address, holder types.Address, options *types.TokenQueryOptions) (map[uint64]*big.Int, error) {
```

Comment:
I noticed that GetERC20Balance was failing in a part of the UI. Right now when we get all the account holders, we made subsequent requests for the balance of each with start and end block both set to the latest block. This is because we want the current balance for each holder as of this block. This broke when i tried the in-memory db, and made me notice that the elasticsearch implementation is broken in a different way.

Let's say there was a token transfer at block 40 and we're on block 45. If we request the balance with starting block 45 and ending block 45, we get:
- in-memory: no result, since the balance wasn't changed in block 45, but rather block 40
- elasticsearch: array of all balance changes, it seems to be ignoring the start block value completely (i thought it was working before because I was only grabbing the first entry in the result array)

I'm not sure how exactly the best way to implement this is. I could see the argument that i should be requesting (1, 45) instead of (45, 45). So I can change it to that if we need to. But I could also see that if i want to know the balance of a holder at block X, I should be able to just request the balance at X without thinking about whether the range i'm giving includes the latest change to that balance.

What do you guys think?

Reply:
I have fixed the bug in both in-memory db and elasticsearch now. If there is no balance at or after `beginBlockNumber` then it would return a single balance which is the last balance change(for the max blockNumber) with blockNumber as `beginBlockNumber`.